Parser/asdl_c.py

```diff
@@ -886,14 +1043,75 @@ def visitModule(self, mod):
         return NULL;
     }
 
-    PyObject *dict;
+    PyObject *dict = NULL, *fields = NULL, *remaining_fields = NULL,
+             *remaining_dict = NULL, *positional_args = NULL;
     if (PyObject_GetOptionalAttr(self, state->__dict__, &dict) < 0) {
         return NULL;
     }
+    PyObject *result = NULL;
     if (dict) {
-        return Py_BuildValue("O()N", Py_TYPE(self), dict);
+        // Serialize the fields as positional args if possible, because if we
+        // serialize them as a dict, during unpickling they are set only *after*
+        // the object is constructed, which will now trigger a DeprecationWarning
+        // if the AST type has required fields.
```

Comment:
Why do AST types even need their own custom reduce function? Shouldn't `object.__reduce__` (which IIRC uses `__new__` to safely create an empty instance, then updates its `__dict__`) work fine for AST objects as well?

Reply:
The `__new__` call will trigger DeprecationWarnings. If we don't have our own reducer, unpickling will essentially do `node = ast.FunctionDef(); node.__dict__.update(...)`, and the first line raises warnings because we don't set the required `name` field.

Reply:
Why would the `__new__` call raise deprecation warnings? Aren't the deprecation warnings in `__init__`, not `__new__`? That's why `object.reduce()` uses `__new__`, so that it can bypass whatever is happening in custom `__init__` methods (otherwise lots of custom classes would fail to unpickle by default). So `object.reduce()` is not equivalent to `node = ast.FunctionDef(); node.__dict__.update(...)` -- it's more like `node = ast.FunctionDef.__new__(ast.FunctionDef); node.__dict__.update(...)`, which is a critical difference. So it seems to me like using the default reducer should work fine here, and not require all this extra work. But I could definitely be missing something; there's probably some reason these nodes had a custom reduce() in the first place.

Reply:
I see what you mean. I tried commenting out the `__reduce__` definition, and lots of tests fail with:

```
ERROR: test_pickling (test.test_ast.AST_Tests.test_pickling) (ast=<ast.Module object at 0x102323da0>, protocol=1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jelle/py/cpython/Lib/test/test_ast.py", line 977, in test_pickling
    ast2 = pickle.loads(pickle.dumps(ast, protocol))
                        ~~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/Users/jelle/py/cpython/Lib/copyreg.py", line 71, in _reduce_ex
    state = base(self)
            ~~~~^^^^^^
TypeError: AST constructor takes at most 0 positional arguments
```

I don't entirely understand what `_reduce_ex` is doing there, but it picks `ast.AST` as the base, which fails.

Reply:
Interesting. Well, I might have just nerd-sniped myself into digging into this more later to understand what's happening, but I don't think it should block the PR; your added code to use positional args works fine.